contributors/design-proposals/ipvs-proxy.md

```diff
@@ -0,0 +1,152 @@
+
+##  Kubernetes behavior change
+
+### Changes to kube-proxy startup parameter
+
+In addition to existing userspace and iptables mode, ipvs mode is configured via `--proxy-mode=ipvs`. In alpha version, it implicitly uses IPVS NAT(masq) mode.
+
+A new kube-proxy parameter `ipvs-scheduler` will be added to specify IPVS load balancing algorithm, the parameter is --ipvs-scheduler.  Below is list of supported values, if it’s not configured `rr` is default value, if it’s incorrectly configured kube-proxy will exit with error message.
+
+- rr: round-robin
+
+
+- lc: least connection
+
+
+- dh: destination hashing
+
+
+- sh: source hashing
+
+
+- sed: shortest expected delay
+
+
+- nq: never queue
+
+For more details about it, refer to [http://kb.linuxvirtualserver.org/wiki/Ipvsadm](http://kb.linuxvirtualserver.org/wiki/Ipvsadm)
+
+In order to clean up inactive rules(includes iptables rules and ipvs rules), we will introduce a new  kube-proxy parameter `--cleanup-proxyrules ` and mark the older `--cleanup-iptables` deprecated. Unfortunately, there is no way to distinguish whether an ipvs service is created by ipvs proxy or other process, `--cleanup-proxyrules` will clear all ipvs service in a host.
+
+### Change to build
+
+No changes. The IPVS implementation relies on [docker/libnetwork](https://godoc.org/github.com/docker/libnetwork/ipvs) ipvs library, which is a pure-golang implementation. 
+
+### Change to deployment
+
+IPVS kernel module installation is beyond kubernetes, it’s assumed IPVS is installed before running kubernetes.  When kube-proxy starts, if proxy mode is IPVS kube-proxy would validate if IPVS is installed on the node, if it’s not installed kube-proxy will exit with an error message.   
+
+Some of the existing deployment script might need update to install ipvsadm, which is beyond the scope of alpha version.
+
+```shell
+apt-get install ipvsadm
+```
+
+## Other design considerations
+
+### IPVS setup and network topology
+
+IPVS is a replacement or IPTables as load balancer, it’s assumed reader of this proposal is familiar with IPTables load balancer mode. We will create a dummy interface and assign all service Cluster IPs to the dummy interface(maybe called `kube0`). In alpha version, we will implicitly use NAT mode. 
+
+We will create some ipvs services for each kubernetes service. The VIP of ipvs service corresponding to the accessable IP(such as cluster IP, external IP, nodeIP, ingress IP, etc.) of kubernetes service. Each destination of an ipvs service corresponding to an kubernetes service endpoint.
+
+### Port remapping 
+
+There are 3 proxy mode in ipvs, NAT, IPIP and DR. Only NAT mode support port remapping.
+
+### NodePort type service support
+
+For NodePort type service, IPVS proxy will take all accessable IPs in a host as the virtual IP of ipvs service. Specifically, accessable IP excludes `lo`, `docker0`, `vethxxx`, `cni0`, `flannel0`, etc. Currently, we assume they are IPs bound to `eth{i}`.
```

Comment:
What about take all UP state network interface(except the "vethxxx"s) address as the Node IP?